sdk/storage/Azure.Storage.Common/src/Shared/SasExtensions.cs

```diff
@@ -288,5 +288,44 @@ internal static void AppendProperties(this SasQueryParameters parameters, String
                 stringBuilder.AppendQueryParameter(Constants.Sas.Parameters.Signature, WebUtility.UrlEncode(parameters.Signature));
             }
         }
+
+        internal static string ValidateAndSanitizeRawPermissions(string permissions,
+            List<char> validPermissionsInOrder)
+        {
+            if (permissions == null)
+            {
+                return null;
+            }
+
+            // Convert permissions string to lower case.
+            permissions = permissions.ToLowerInvariant();
+
+            HashSet<char> validPermissionsSet = new HashSet<char>(validPermissionsInOrder);
+            HashSet<char> permissionsSet = new HashSet<char>();
+
+            foreach (char permission in permissions)
+            {
+                // Check that each permission is a real SAS permission.
+                if (!validPermissionsSet.Contains(permission))
+                {
+                    throw new ArgumentException($"{permission} is not a valid SAS permission");
```

Comment:
This is a nontrivial change.  We've always allowed customers to do whatever they want with the raw permissions, but now you're forcing them to be a set of known values.  Why not let the service reject unknown values?  Why aren't we putting this behind something like an optional `bool normalize = false` param?  Are we doing this consistently across languages?

Reply:
The service doesn't return a useful error message if you include an invalid permission - just "Unauthorized".  The user has no way of knowing that they did wrong.  Also, how do I re-order the permissions if the user provides permissions that don't exist?

@gapra-msft, @xiafu-msft, do we allow users to pass invalid raw SAS permissions in in Java and Python?

Reply:
@tg-msft, my bad, I accidentally edited your comment and then reverted it back.

Reply:
Can we get a bug filed against the service to return a better error message for an unknown permission?  It should be fine to explain why a SAS permission is invalid per the service API rather than relative to a resource.

And yes, you can't order permissions you don't know about which is why I'm suggesting moving this validation + reordering logic behind an optional `normalize` flag.  We normally don't validate anything on clients so we should make it optional if we're providing this as a convenience.  This would also enable @blueww's scenario from above.

+@XiaoningLiu for JS as well.

Reply:
Fixed in next iteration.

Reply:
@tg-msft Java does not allow to pass raw permissions verbatim - they're always validated and ordered before generating sas. Python seems to accept union of *SasPermission (which validates and orders) and raw string, so user can pass raw string without any validation there.

Reply:
So it sounds like Java is the outlier with no option to skip validation and we don't know about JS yet.  Is it worth filing an issue for Java?  I could see waiting to add this until someone asks because it wouldn't be a breaking change.

Reply:
I feel that if I were implementing that from scratch I'd chose Java's approach as it reduces # of ways customer can get it wrong (it takes care of the ordering). So I'd rather change .NET and Python in next major rev. I think we should hold with changing Java until someone explicitly asks for that (not sure why would they want to do that).